File: 2016/03/lib/aocday/parser.ex
defmodule AOCDay.Parser do
  def parse do
    {:ok, binary} = File.read("./input.txt")
    _parse(binary)
  end

  def parse(binary) do
    _parse(binary)
  end

  defp _parse(binary) do
    binary
    |> String.trim
    |> String.split("\n")
    |> Enum.map(fn(line) ->
      line
      |> String.trim
      |> String.split(~r/\s+/)
      |> Enum.map(&(String.to_integer/1))
      |> Enum.map(&(abs/1))
    end)
  end
end

File: 2020/7/lib/parser.rb
# frozen_string_literal: true

class Parser
  def self.parse(input = nil)
    input ||= File.read(File.join(File.dirname(__FILE__), "../input.txt"))

    input
      .strip
      .split("\n")
      .each_with_object({}) { |l, acc| parse_line(l, acc) }
  end

  def self.parse_line(line, acc)
    outer, contained = line.split("bags contain").map(&:strip)
    acc[outer] = contained
      .split(",")
      .map { |l| l.match(/(?<number>\d+) (?<bag>[\w\s]+) bag/) }
      .each_with_object({}) do |match, acc2|
        acc2[match["bag"]] = match["number"].to_i if match
        acc2
      end
  end
end
